### PR TITLE
[FIX] Make sure Python >=3.9 is supported

### LINKED ISSUE #236: Self import from typing does not work with Python 3.10
```from typing import Self```

Does not work with Python 3.10 as it is only available in Python 3.11+. 

This import is directly stated in the following files:

nipoppy_cli/nipoppy/tabular/base.py
nipoppy_cli/nipoppy/tabular/doughnut.py
nipoppy_cli/nipoppy/config/main.py
nipoppy_cli/nipoppy/tabular/manifest.py

A possible option is to use 

```from typing_extensions import Self```

 instead, as mentioned in this [stack overflow solution](https://stackoverflow.com/a/77247460).

Or mention in the Docs to use Python 3.11+ instead of Python 3.9+.

### PR DESCRIPTION
**Pull request by michellewang**
Closes #236

- Update CI workflow for tests
- Update `pyproject.toml` to 
- Add `from __future__ import annotations` to files that use `|` for `typing.Union` (only available in Python >= 3.10 otherwise)
- Use `from typing_extensions import Self` instead of `from typing import Self` (only available in Python >= 3.11) 
- Combine `str | pathlib.Path` into `StrOrPathLike` typevar

### CODE REVIEW

File: nipoppy_cli/nipoppy/workflows/dicom_reorg.py [hunk 1]
```diff
@@ -13,7 +14,7 @@ class DicomReorgWorkflow(BaseWorkflow):
 
     def __init__(
         self,
-        dpath_root: Path | str,
+        dpath_root: StrOrPathLike,
         copy_files: bool = False,
         fpath_layout: Optional[Path] = None,
```
Comment (nikhil153) [inline nipoppy_cli/nipoppy/workflows/dicom_reorg.py:19]: in `pipeline.py` this is set as: `fpath_layout: Optional[StrOrPathLike] = None,` 
But several other places it's still `fpath_layout: Optional[Path] = None`

**Review by nikhil153 — APPROVED** [review]
Just a couple of minor comments.

File: nipoppy_cli/nipoppy/workflows/dicom_reorg.py (hunk 1 above)
Comment (michellewang) [reply nipoppy_cli/nipoppy/workflows/dicom_reorg.py]: Good catch, fixed!

File: nipoppy_cli/nipoppy/workflows/base.py [hunk 2]
```diff
@@ -101,7 +103,7 @@ def run_command(
 
         Parameters
         ----------
-        command_or_args : Sequence[str] | str
+        command_or_args : Sequence[str]  |  str
```
[the remark this answers is not in the document]
Comment (michellewang) [reply nipoppy_cli/nipoppy/workflows/base.py:106]: This will affect how the autodocs look but I don't think that matters (for now at least)

**Pull request merged**